src/back/msl/writer.rs

```diff
@@ -1721,9 +1721,56 @@ impl<W: Write> Writer<W> {
                     self.put_expression(arg, context, true)?;
                     write!(self.out, ") + 1) % 33) - 1)")?;
                 } else if fun == Mf::FindMsb {
-                    write!(self.out, "((({NAMESPACE}::clz(")?;
-                    self.put_expression(arg, context, true)?;
-                    write!(self.out, ") + 1) % 33) - 1)")?
+                    let inner = context.resolve_type(arg);
+
+                    if let Some(crate::ScalarKind::Uint) = inner.scalar_kind() {
+                        write!(self.out, "((({NAMESPACE}::clz(")?;
+                        self.put_expression(arg, context, true)?;
+                        write!(self.out, ") + 1) % 33) - 1)")?;
+                    }
+
+                    match *inner {
+                        crate::TypeInner::Vector {
+                            size,
+                            kind: crate::ScalarKind::Sint,
+                            ..
+                        } => {
+                            let s = back::vector_size_str(size);
+                            write!(self.out, "{NAMESPACE}::select(int{s}(")?;
+                            write!(self.out, "{NAMESPACE}::floor(")?;
+                            write!(self.out, "{NAMESPACE}::log2(float{s}(")?;
```

Comment:
I would expect using a floating-point logarithm here to be much slower than using `metal::clz`. Could you see if that can work?

Reply:
Yeah small oversight on my part, I can fix this.